onnxruntime/core/providers/webnn/builders/impl/einsum_op_builder.cc

```diff
@@ -0,0 +1,598 @@
+
+  auto equals = [](gsl::span<const uint32_t> a, gsl::span<const uint32_t> b) {
+    return std::equal(a.begin(), a.end(), b.begin(), b.end());
+  };
+
+  std::array<uint32_t, 3> component_ranks;
+  if (m_components.size() > component_ranks.size()) {
+    // So far, not support for more than two inputs and one output.
+    return RecognizedOperatorType::None;
+  } else if (m_components.size() == 2) {  // one input
+    auto input_labels = m_components[0].GetLabels(m_label_indices);
+    auto output_labels = m_components[1].GetLabels(m_label_indices);
+    if (input_labels.size() == output_labels.size()) {
+      if (equals(input_labels, output_labels)) {  // identity
+        return RecognizedOperatorType::Identity;
+      } else {
+        return RecognizedOperatorType::Transpose;
+      }
+    } else if (output_labels.empty()) {  // scalar output, reduce
+      return RecognizedOperatorType::ReduceSum;
+    }
+
+  } else if (m_components.size() == 3) {  // two inputs
+    auto input_A_labels = m_components[0].GetLabels(m_label_indices);
+    auto input_B_labels = m_components[1].GetLabels(m_label_indices);
+    auto output_labels = m_components[2].GetLabels(m_label_indices);
+    if (equals(input_A_labels, output_labels) && equals(input_B_labels, output_labels)) {  // element-wise product
+      return RecognizedOperatorType::Multiply;
+    }
+  }
+
+  return RecognizedOperatorType::Others;
+}
+
+// Add operator related.
+
+Status EinsumOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder,
+                                              const Node& node,
+                                              const logging::Logger& logger) const {
+  const auto& input_defs = node.InputDefs();
+  const auto& initializers(model_builder.GetInitializerTensors());
+  emscripten::val output = emscripten::val::object();
+
+  NodeAttrHelper helper(node);
+  const auto equation = helper.Get("equation", std::string(" "));
+
+  std::vector<uint32_t> m_label_indices;
+  std::vector<Component> m_components;
+  std::vector<uint32_t> m_output_dimensions;
+  uint32_t num_labels;
+  ORT_RETURN_IF_NOT(ParseEquationComponents(initializers, node, equation, m_label_indices,
+                                            m_components, m_output_dimensions, num_labels, logger),
+                    "Error parsing equation components.");
+
+  RecognizedOperatorType recognized_operator_type = DetermineRecognizedOperatorType(m_label_indices, m_components,
+                                                                                    m_output_dimensions);
+
+  switch (recognized_operator_type) {
+    case RecognizedOperatorType::Multiply: {
+      const size_t a_idx = 0, b_idx = 1;
+      emscripten::val a = model_builder.GetOperand(node.InputDefs()[a_idx]->Name());
+      emscripten::val b = model_builder.GetOperand(node.InputDefs()[b_idx]->Name());
+      output = model_builder.GetBuilder().call<emscripten::val>("mul", a, b);
+    } break;
```

Comment:
🔎👀 Dense code is tougher to read. Inserting some line breaks between the case statements would help readers logically see the separate blocks better.